math/src/main/kotlin/ktx/math/ImmutableVector.kt

```diff
@@ -0,0 +1,187 @@
+@file:Suppress("NOTHING_TO_INLINE")
+
+package ktx.math
+
+import com.badlogic.gdx.math.Interpolation
+import com.badlogic.gdx.math.MathUtils
+import com.badlogic.gdx.math.Vector
+import kotlin.math.abs
+import kotlin.math.sqrt
+
+/**
+ * Represent an immutable vector.
+ *
+ * This is the root interface of immutable alternatives to the default mutable vectors provided by LibDGX
+ */
+interface ImmutableVector<T : ImmutableVector<T>> : Comparable<T> {
+
+    /**
+     * Returns the squared euclidean length
+     *
+     * This method is faster than [Vector.len] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact lengths, as the return value is the square of the actual length.
+     */
+    val len2: Float
+
+    /** Returns the euclidean length */
+    val len: Float get() = sqrt(len2)
+
+    /**
+     * Returns the unit vector of same direction or this vector if it is zero.
+     */
+    val nor: T
+
+    /** Returns whether the length of this vector is smaller than the given [margin] */
+    fun isZero(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Returns whether this vector is a unit length vector within the given [margin]. (no margin by default) */
+    fun isUnit(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean = abs(1f - len2) < margin
+
+    /** Returns the opposite vector of same length */
+    operator fun unaryMinus(): T
+
+    /** Returns the result of subtracting the [other] vector from this vector */
+    operator fun minus(other: T): T
+
+    /** Returns the result of adding the [other] vector to this vector */
+    operator fun plus(other: T): T
+
+    /** Returns a new vector instance with all members incremented by 1 */
+    operator fun inc(): T
+
+    /** Returns a new vector instance with all members decremented by 1 */
+    operator fun dec(): T
+
+    /** Returns a new vector instance scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** Returns a new vector instance scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** Returns a vector of the same direction and a squared length of [length2] */
+    fun withLength2(length2: Float): T
+
+    /** Returns the dot product of this vector by the given [vector] */
+    infix fun dot(vector: T): Float
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    infix fun dst2(vector: T): Float
+
+    /** Linearly interpolates between this vector and the target vector by alpha */
```

Comment:
It seems that the current Kotlin convention proposes to include `[references]` to parameters in the doc comments. Using `@param` would be excessive, but I'd appreciate if you wrapped at least one reference to each parameter mentioned in each doc comment with a link. If you don't have the time, this is OK - it's not crucial, I can add it after merging.